python/singa/autograd.py

```diff
@@ -3239,176 +3238,152 @@ def __init__(self,
-                                         self.Bh)
-                out.append(h)
-            return out, h, c
-        elif self.backend == "cudnn":
-            if not singa.USE_CUDA:
-                raise Exception("Could not use cudnn without cuda compiled.\n")
+        return out, h, c
 
-            cpp_x = singa.VecTensor()
-            [cpp_x.append(i.data) for i in xs]
+    def step_forward(self, x, h, c, Wx, Wh, Bx, Bh):
+        y1 = matmul(x, Wx[0])
+        y1 = add_bias(y1, Bx[0], axis=0)
+        y2 = matmul(h, Wh[0])
+        y2 = add_bias(y2, Bh[0], axis=0)
+        i = add(y1, y2)
+        i = sigmoid(i)
+
+        y1 = matmul(x, Wx[1])
+        y1 = add_bias(y1, Bx[1], axis=0)
+        y2 = matmul(h, Wh[1])
+        y2 = add_bias(y2, Bh[1], axis=0)
+        f = add(y1, y2)
+        f = sigmoid(f)
+
+        y1 = matmul(x, Wx[2])
+        y1 = add_bias(y1, Bx[2], axis=0)
+        y2 = matmul(h, Wh[2])
+        y2 = add_bias(y2, Bh[2], axis=0)
+        o = add(y1, y2)
+        o = sigmoid(o)
+
+        y1 = matmul(x, Wx[3])
+        y1 = add_bias(y1, Bx[3], axis=0)
+        y2 = matmul(h, Wh[3])
+        y2 = add_bias(y2, Bh[3], axis=0)
+        g = add(y1, y2)
+        g = tanh(g)
+
+        cout1 = mul(f, c)
+        cout2 = mul(i, g)
+        cout = add(cout1, cout2)
+
+        hout = tanh(cout)
+        hout = mul(o, hout)
+        return hout, cout
+
+
+class _RNN_cudnn(Operation):
+    def __init__(self, handle):
+        super(_RNN_cudnn, self).__init__()
+        self.handle = handle
 
-            self.W = Tensor(shape=(self.rnn_handle.weights_size,),
-                            requires_grad=True,
-                            stores_grad=True)
-            self.W.gaussian(0.0, 1.0)
+    def forward(self, x, W):
+        y = singa.GpuRNNForwardTraining(x, W, self.handle)
+        if training:
+            self.inputs = (x, W, y)
+        return y
 
-            cpp_y = singa.GpuRNNForwardTraining(cpp_x, self.W.data,
-                                                self.rnn_handle)
-            y = self.cpp_vec_tensor_to_py_tensor(cpp_y)
+    def backward(self, dy):
+        assert training is True and hasattr(
+            self, "inputs"), "Please set training as True before do BP. "
 
-            if training:
-                self.buffer = {"cpp_y": cpp_y, "cpp_x": cpp_x}
+        dx = singa.GpuRNNBackwardx(self.inputs[2], dy, self.inputs[1], self.handle)
+        dW = singa.GpuRNNBackwardW(self.inputs[0], self.inputs[2], self.handle)
+        return dx, dW
 
-            return y
-        else:
-            raise Exception("Unsupported backend %s.\n" % backend)
+class RNN_cudnn(Layer):
+    def __init__(self, input_size, hidden_size, rnn_mode=2):
+        """
+            Args:
+                input_size: input feature dim
+                hidden_size: hidden feature dim
+                rnn_mode: could be 0 - RNN RELU, 1 - RNN TANH, 2 - LSTM, 3 - GRU
```

Comment:
use string, 'relu', 'tanh', gru'.
does it support vanilla rnn with 'tanh' ?

Reply:
this is resolved